site/blog/unlisted/why-sql-for-infrastructure.mdx

```diff
@@ -0,0 +1,94 @@
+- By defining [trigger functions](https://www.postgresql.org/docs/current/sql-createtrigger.html#SQL-CREATETRIGGER-EXAMPLES) an [audit log](https://en.wikipedia.org/wiki/Audit_trail) can be maintained, which can keep track of who made what cloud changes and when they did.
+- Other trigger functions can be added to automatically re-insert deleted records for automated [Chaos Engineering](https://en.wikipedia.org/wiki/Chaos_engineering) recovery.
+- [Snapshotting](https://en.wikipedia.org/wiki/Snapshot_%28computer_storage%29) could be used to quickly revert all infrastructure changes without hidden state (eg, the configuration of your load balancers, but not the data within your own databases).
+- By having a standardized protocol to interface with the database, you can:
+  - Perform operations on it via scripts and the [`psql` CLI](https://www.postgresql.org/docs/current/app-psql.html) tool.
+  - Integrate it into your application itself with a [postgres client library](https://node-postgres.com/) for application-driven infrastructure management (like provisioning sharded resources for a client that wants isolation).
+  - Connect to it with a [SQL IDE](https://www.jetbrains.com/datagrip/features/postgresql/) for outage mitigation or inspection in a tabular (Excel-like) view.
+  - Connect it to a [graphical dashboard](https://grafana.com/docs/grafana/latest/datasources/postgres/) for live visualization of your infrastructure.
+  - Connect it to a [low code tool](https://docs.retool.com/docs/postgresql-integration) for rapid response dashboards for your infrastructure with buttons you can tackle known outage vectors near-instantly.
+
+## SQL is the best option for Infrastructure Management
+
+There are negatives to representing your infrastructure as SQL, of course, as there is no perfect solution for all situations, but we believe IaSQL is a superset versus IaC in all but one category.
+
+- SQL is an old, irregular language to work with, but it is better known than HCL and we already have a Pulumi in the form of every ORM (like [TypeORM](https://typeorm.io/), [ActiveRecord](https://guides.rubyonrails.org/active_record_basics.html), etc) and QueryBuilder ([LINQ](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/), [Knex](https://knexjs.org/), etc) in whatever programming language you prefer.
+- Peer review of changes is *different* versus IaC. [Migration tools](https://en.wikipedia.org/wiki/Schema_migration) *can* be used to approximate the IaC review process, with the CI system wrapping the migration SQL in an [IaSQL transaction](/docs/transaction) and previewing the operation, but since IaSQL transactions are intentionally *not* actual Postgres transactions, you can simply make the proposed infrastructure changes in the database within an uncommitted IaSQL transaction and allow your peers to inspect/review the database state and call the commit, with the audit log keeping track of the change history for you.
```

Comment:
I would really like to remove this. We have learned the hard way that we don't want to recommend migration tools after attempting to use them ourselves and failing multiple times. I would call out and prioritize the idea of creating a GitHub module that makes read-only side-effect PRs that can be called alongside transactions that push preview output and sql from audit table to a markdown file that teammates can approve. We can link to the task in the post